### ISSUE TITLE
ros service - no module named `yaml`

### ISSUE THREAD

**Issue by i-am-logger** [lopsided98/nix-ros-overlay#367]
I modified the talker from the demo cpp to a heartbeat service and i get the following error:

![image](https://github.com/lopsided98/nix-ros-overlay/assets/1440852/a0a834db-1bc9-49ba-8801-5cad0fe059f2)

my fle: 

```nix
 inputs = {
    nixpkgs.url = "nixpkgs/nixos-unstable";
    nixos-rk3588 = {
      url = "github:realsnick/nixos-rk3588";
    };
    nix-ros-overlay = {
      url = "github:lopsided98/nix-ros-overlay";
      inputs.nixpkgs.follows = "nixpkgs";
    };
  };
  ...
  sky360-packages-overlay = (final: prev: {
    sky360 = {
      sky360-ros-iron-heartbeat = prev.callPackage ./ros/iron/heartbeat { };
      qhyccd_sdk = prev.callPackage ./pkgs/qhyccd_sdk.nix { };
    };
  });
... 
lib.nixosSystem {
  inherit system;
    pkgs = import nixpkgs {
      inherit system;
      config.allowUnfree = true;
      overlays = [
        nix-ros-overlay.overlays.default
        sky360-packages-overlay
      ];
    };
  };
```
ros2.nix
```nix
{ pkgs, ... }:
{
  environment.systemPackages = with pkgs; [
    sky360.sky360-ros-iron-heartbeat
  ];

  services.ros2 = {
    enable = true;
    distro = "iron";
    domainId = 0;
    nodes = {
      heartbeat = {
        package = pkgs.sky360.sky360-ros-iron-heartbeat.pname;
        node = "heartbeat";
        args = [ ];
        rosArgs = [ ];
        params = { };
      };
    };
  };
}
```
heartbeat.nix

```nix
{ pkgs, lib, ... }:
with pkgs;
with pkgs.rosPackages.iron;

buildRosPackage {
  pname = "sky360-ros-iron-heartbeat";

  version = "0.1.1";

  src = ./ros;

  buildType = "ament_cmake";
  buildInputs = [ ament-cmake rmw-implementation-cmake ];
  checkInputs = [ ament-cmake-pytest ament-lint-auto ament-lint-common launch launch-testing launch-testing-ament-cmake launch-testing-ros ];
  propagatedBuildInputs = [ example-interfaces launch-ros launch-xml rcl-interfaces rclcpp rclcpp-components rcutils rmw std-msgs ];
  nativeBuildInputs = [ ament-cmake ];

  meta = {
    description = ''Sky360 heartbeat ros2 node'';
    license = with lib.licenses; [ mit ];
  };
}
```

**Comment by wentasah** [comment]
I cannot easily reproduce your problem with the files you provided (few things are missing there), but it appears that [`package.xml` in `rclpy`](https://github.com/ros2/rclpy/blob/rolling/rclpy/package.xml) lacks the following line:

    <depend>python3-yaml</depend>

As a quick workaround, you can add `python3Packages.pyyaml` to `propagatedBuildInputs` in your `heartbeat.nix`.

To fix this problem for everybody, I suggest to first test whether adding `python3Packages.pyyaml` to `propagatedBuildInputs` in `distros/iron/rclpy/default.nix` (and adding `python3Packages` to that function parameters) fixes the problem. If yes, than it would be great if you can send a PR adding the above `<depend>` line to `rclpy`. Mention there the error message you're experiencing. After some time, the fix will automatically appear in this overlay too.

**Comment by realsnick** [comment]
tried to add `python3Packages.pyyaml` but still fails with same error :/

**Comment by wentasah** [comment]
Hmm, right. I verified that `rospy` has the `yaml` module available, perhaps through some of its dependencies. So the problem is elsewhere. I tried clonning your repo https://github.com/realsnick/sky360/ and running:

    nixos-rebuild build-vm --flake .#cyclop-x86_64

But it fails with:

```
error: The option `sdImage' does not exist. Definition values:
       - In `/nix/store/cxlqv54wad2n39g1axz8k9pra006d80h-source/nixOS/flake.nix':
           {
             imageName = "sky360-cyclop-0.3.2-x86_64-linux";
           }
```
Do you know why?

**Comment by realsnick** [comment]
```nix build --option system aarch64 .#cyclop-orange_pi_5_plus-sd_image``` is the command

**Comment by realsnick** [comment]
i made a mistake! apologizes...adding the `pyyaml` to `rcply` seems to fix one issue but now it seems the service fails due to unable to find the package

![image](https://github.com/lopsided98/nix-ros-overlay/assets/1440852/7b74feb0-54dc-4fec-8818-fefd382685b2)

**Comment by wentasah** [comment]
I have no practical experience with NixOS ROS modules, but from looking at the code, it seems you should modify your `ros2.nix` as follows:
```nix
{ pkgs, ... }:
{
  services.ros2 = {
    enable = true;
    distro = "iron";
    domainId = 0;
    inherit pkgs;
    systemPackages = pkgs: with pkgs; [
      sky360.sky360-ros-iron-heartbeat
    ];
    nodes = {
      heartbeat = {
        package = pkgs.sky360.sky360-ros-iron-heartbeat.pname;
        node = "heartbeat";
        args = [ ];
        rosArgs = [ ];
        params = { };
      };
    };
  };
}
```

**Comment by realsnick** [comment]
i tried that:

the `inherit pkgs` is the issue here

![image](https://github.com/lopsided98/nix-ros-overlay/assets/1440852/207205a9-bba7-411c-93b8-c232d30ac9b9)

**Comment by wentasah** [comment]
I'd temporarily modify `modules/ros2/nodes.nix` to print the `$AMENT_PREFIX_PATH` variable before starting `ros2 run` (or `ros2 launch`) and look manually, whether the package is really there and whether it has all needed ROS metadata.

**Comment by realsnick** [comment]
seems this variable isn't defined... i ran `/bin/sh -c env`

![image](https://github.com/lopsided98/nix-ros-overlay/assets/1440852/caa5bf58-4a47-4ba2-aaab-c1049b97f64c)

**Comment by wentasah** [comment]
Yeah, this seems tricky, but I may have some clue what's going on. `nodes.nix` uses ROS specific `buildEnv` to construct the environment with the needed packages. By default, `buildEnv` works by wrapping all binaries in `.../bin` to set variables like `$AMENT_PREFIX_PATH`. So in order to start the service correctly, the `ros2` command (which is used in the systemd service files) must be in the same `buildEnv` as the node package. If I read `nodes.nix` correctly, it attempts to do it, but there might be bugs.

Can you verify that the `ros2` command started by the service (`ExecStart` in your very first screenshot above (BTW sending the text wrapped in ` ``` ` would be easier to work with that the screenshots)) is actually a wrapper, which sets `AMENT_PREFIX_PATH`? Then look into the paths mentioned there and see if your package is in `share/` subdirectory of those paths.

**Comment by realsnick** [comment]
i'll take a deeper look , thank you!

have some errands this weekend but will get to it as soon as i can...
